<a href="https://colab.research.google.com/github/vietphtran/Data/blob/master/Exploring_Ebay_Car_Sales_Data.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# **Exploring Ebay Car Sales Data**

In this project, we, a team of three data analysts, work with a dataset of used cars being sold on [eBay Kleinanzeigen](https://www.ebay-kleinanzeigen.de/), the German version of eBay. The dataset was sampled from the original dataset on Kaggle and consists of 50,000 data points.

Our main goal is to clean and analyze to provide useful insights regarding used cars being listed on German eBay website.

## **Import and Explore the data set**

We will primarily work with the data set using two Python libraries: numpy and pandas. Our first step is to import these libraries as well as our dataset ([download here](https://github.com/vietphtran/Data/raw/master/autos.csv)).

In [1]:
import numpy as np
import pandas as pd

In [2]:
path = "https://github.com/vietphtran/Data/raw/master/autos.csv"
autos = pd.read_csv(path, encoding="Latin-1")

In [3]:
autos

dateCrawled  ...             lastSeen
0      2016-03-26 17:47:46  ...  2016-04-06 06:45:54
1      2016-04-04 13:38:56  ...  2016-04-06 14:45:08
2      2016-03-26 18:57:24  ...  2016-04-06 20:15:37
3      2016-03-12 16:58:10  ...  2016-03-15 03:16:28
4      2016-04-01 14:38:50  ...  2016-04-01 14:38:50
...                    ...  ...                  ...
49995  2016-03-27 14:38:19  ...  2016-04-01 13:47:40
49996  2016-03-28 10:50:25  ...  2016-04-02 14:18:02
49997  2016-04-02 14:44:48  ...  2016-04-04 11:47:27
49998  2016-03-08 19:25:42  ...  2016-04-05 16:45:07
49999  2016-03-14 00:42:12  ...  2016-04-06 21:18:48

[50000 rows x 20 columns]

As mentioned in the introduction, the dataset consists of 50,000 rows and 20 columns. The data dictionary provided with data is as follows:

- **dateCrawled** - When this ad was first crawled. All field-values are taken from this date.
- **name** - Name of the car.
- **seller** - Whether the seller is private or a dealer.
- **offerType** - The type of listing
- **price** - The price on the ad to sell the car.
- **abtest** - Whether the listing is included in an A/B test.
- **vehicleType** - The vehicle Type.
- **yearOfRegistration** - The year in which the car was first registered.
- **gearbox** - The transmission type.
- **powerPS** - The power of the car in PS.
- **model** - The car model name.
- **kilometer** - How many kilometers the car has driven.
- **monthOfRegistration** - The month in which the car was first registered.
- **fuelType** - What type of fuel the car uses.
- **brand** - The brand of the car.
- **notRepairedDamage** - If the car has a damage which is not yet repaired.
- **dateCreated** - The date on which the eBay listing was created.
- **nrOfPictures** - The number of pictures in the ad.
- **postalCode** - The postal code for the location of the vehicle.
- **lastSeenOnline** - When the crawler saw this ad last online.


Below, we print some basic summaries regardign each column in the dataset. Some initial observations:
- Column names are camelcase and should be changed to snakecase to avoid unnecessary typos and errors;
- 5 columns contain numeric values and 15 columns contain string values. However, taking a closer look, we can see that columns like price and odometer should have numeric values instead. Thus, we expect some data cleaning and manipulation for some columns including these two;
- 5 columns have missing values that we should investigate as well progress through the project;
- Some columns have German words instead of English.


In [4]:
autos.info()
autos.head()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 20 columns):
 #   Column               Non-Null Count  Dtype 
---  ------               --------------  ----- 
 0   dateCrawled          50000 non-null  object
 1   name                 50000 non-null  object
 2   seller               50000 non-null  object
 3   offerType            50000 non-null  object
 4   price                50000 non-null  object
 5   abtest               50000 non-null  object
 6   vehicleType          44905 non-null  object
 7   yearOfRegistration   50000 non-null  int64 
 8   gearbox              47320 non-null  object
 9   powerPS              50000 non-null  int64 
 10  model                47242 non-null  object
 11  odometer             50000 non-null  object
 12  monthOfRegistration  50000 non-null  int64 
 13  fuelType             45518 non-null  object
 14  brand                50000 non-null  object
 15  notRepairedDamage    40171 non-null  object
 16  date

dateCrawled  ...             lastSeen
0  2016-03-26 17:47:46  ...  2016-04-06 06:45:54
1  2016-04-04 13:38:56  ...  2016-04-06 14:45:08
2  2016-03-26 18:57:24  ...  2016-04-06 20:15:37
3  2016-03-12 16:58:10  ...  2016-03-15 03:16:28
4  2016-04-01 14:38:50  ...  2016-04-01 14:38:50

[5 rows x 20 columns]

## **Rename column names: camelcase to snakecase**

We decide to change column names from camelcase to snakecase to make them easier to read and avoid unnecessary typos and mistakes due to the capitalized letters. To do so, we first print out the original column names, then assign the following fixes:

- yearOfRegistration to registration_year
- monthOfRegistration to registration_month
- notRepairedDamage to unrepaired_damage
- dateCreated to ad_created
- The rest of the columnn names from camelcase to snakecase.


In [5]:
autos.columns

Index(['dateCrawled', 'name', 'seller', 'offerType', 'price', 'abtest',
       'vehicleType', 'yearOfRegistration', 'gearbox', 'powerPS', 'model',
       'odometer', 'monthOfRegistration', 'fuelType', 'brand',
       'notRepairedDamage', 'dateCreated', 'nrOfPictures', 'postalCode',
       'lastSeen'],
      dtype='object')

In [6]:
autos.columns = ['date_crawled', 'name', 'seller', 'offer_type', 'price', 'abtest',
       'vehicle_type', 'registration_year', 'gearbox', 'power_ps', 'model',
       'odometer', 'registration_month', 'fuel_type', 'brand',
       'unrepaired_damage', 'ad_created', 'nr_of_pictures', 'postal_code',
       'last_seen']
autos.head()

date_crawled  ...            last_seen
0  2016-03-26 17:47:46  ...  2016-04-06 06:45:54
1  2016-04-04 13:38:56  ...  2016-04-06 14:45:08
2  2016-03-26 18:57:24  ...  2016-04-06 20:15:37
3  2016-03-12 16:58:10  ...  2016-03-15 03:16:28
4  2016-04-01 14:38:50  ...  2016-04-01 14:38:50

[5 rows x 20 columns]

## **Some further observations**

In order to decide the next cleaning tasks, we need to take a closer look at the data (now with easier-to-read columns!). The following table present some descriptive statistics of the data set.

In [7]:
autos.describe(include='all')

date_crawled         name  ...   postal_code            last_seen
count                 50000        50000  ...  50000.000000                50000
unique                48213        38754  ...           NaN                39481
top     2016-03-19 17:36:18  Ford_Fiesta  ...           NaN  2016-04-07 06:17:27
freq                      3           78  ...           NaN                    8
mean                    NaN          NaN  ...  50813.627300                  NaN
std                     NaN          NaN  ...  25779.747957                  NaN
min                     NaN          NaN  ...   1067.000000                  NaN
25%                     NaN          NaN  ...  30451.000000                  NaN
50%                     NaN          NaN  ...  49577.000000                  NaN
75%                     NaN          NaN  ...  71540.000000                  NaN
max                     NaN          NaN  ...  99998.000000                  NaN

[11 rows x 20 columns]

From the above descriptive statistics, some observations can be made as below:

**1. Candidates to be dropped:**
- **nr_of_pictures**: has no values, provide no information;
- **seller** and **offer_type**: only have 2 unique values, values are stored as string and do not seem to provide much information.

To make sure that we won't exclude valuable data, we decide to take a look at the value counts of each columns.

In [8]:
autos["seller"].value_counts()

privat        49999
gewerblich        1
Name: seller, dtype: int64

In [9]:
autos["nr_of_pictures"].value_counts()

0    50000
Name: nr_of_pictures, dtype: int64

In [10]:
autos["offer_type"].value_counts()

Angebot    49999
Gesuch         1
Name: offer_type, dtype: int64

As expected, these columns indeed do not provide much information that is useful for our analysis. Therefore, we decide to exclude these columns from the dataset and moving forward with other observations.

In [11]:
autos.drop(["seller", "nr_of_pictures", "offer_type"], axis=1, inplace=True)
autos.columns

Index(['date_crawled', 'name', 'price', 'abtest', 'vehicle_type',
       'registration_year', 'gearbox', 'power_ps', 'model', 'odometer',
       'registration_month', 'fuel_type', 'brand', 'unrepaired_damage',
       'ad_created', 'postal_code', 'last_seen'],
      dtype='object')

**2. Columns that need more investigation:**
- **registration_year**: the min and max values are suspicous. This column provides information about *The year in which the car was first registered*. Obviously, a car cannot be registered before car was invented (1800s) or in the future.
- **registration_month**: the min value shows 0, which is also suspicous since the month values should range from 1 to 12.

We will take a further look into these columns later on when we work with Dates and Times. For now, we will keep moving on with other observations.

In [12]:
autos["registration_month"].value_counts()

0     5075
3     5071
6     4368
5     4107
4     4102
7     3949
10    3651
12    3447
9     3389
11    3360
1     3282
8     3191
2     3008
Name: registration_month, dtype: int64

In [13]:
autos["registration_year"].value_counts()

2000    3354
2005    3015
1999    3000
2004    2737
2003    2727
        ... 
1931       1
1929       1
1001       1
9996       1
1952       1
Name: registration_year, Length: 97, dtype: int64

**3. Columns with numeric data stored as text**
- price: the values contain special characters (e.g., $5,000)
- odometer: the values contain the unit ('km') and special character (',') (e.g., 5,000km)

We return the correct data types for these columns, we remove such characters and convert the data into numberic values. We also rename these columns as follow:
- price to price_usd
- odometer to odometer_km

These changes are necessary for us to keep track of the unit of these values.

In [14]:
autos["price"] = autos["price"].str.replace("$","").str.replace(",","").astype(int)
autos.rename({"price":"price_usd"}, axis=1, inplace=True)
autos["price_usd"]

0         5000
1         8500
2         8990
3         4350
4         1350
         ...  
49995    24900
49996     1980
49997    13200
49998    22900
49999     1250
Name: price_usd, Length: 50000, dtype: int64

In [15]:
autos["odometer"] = autos["odometer"].str.replace("km","").str.replace(",","").astype(int)
autos.rename({"odometer":"odometer_km"}, axis=1, inplace=True)
autos["odometer_km"]

0        150000
1        150000
2         70000
3         70000
4        150000
          ...  
49995    100000
49996    150000
49997      5000
49998     40000
49999    150000
Name: odometer_km, Length: 50000, dtype: int64

## **Working with price_usd and odometer_usd**

With the two columns now converted to numberic values, we can re-run the descriptive statistics for these columns.

**1. price_usd**

In [16]:
autos["price_usd"].describe()

count    5.000000e+04
mean     9.840044e+03
std      4.811044e+05
min      0.000000e+00
25%      1.100000e+03
50%      2.950000e+03
75%      7.200000e+03
max      1.000000e+08
Name: price_usd, dtype: float64

In [17]:
autos["price_usd"].value_counts().sort_index(ascending=False).head(10)

99999999    1
27322222    1
12345678    3
11111111    2
10000000    1
3890000     1
1300000     1
1234566     1
999999      2
999990      1
Name: price_usd, dtype: int64

The value counts table above of the column price_usd raises some suspicions. For example, some cars are being sold for $100 millions. We don't think these listings are realistic and reasonale, thus we decide to exclude listings with prices higher than 100,000 USD from the dataset.


In [18]:
autos_high = autos[autos["price_usd"] > 100000]
autos_high[["name", "price_usd"]].sort_values("price_usd", ascending=False).head(10)

name  price_usd
39705                        Tausch_gegen_gleichwertiges   99999999
42221                                  Leasinguebernahme   27322222
39377                        Tausche_volvo_v40_gegen_van   12345678
27371                                         Fiat_Punto   12345678
47598  Opel_Vectra_B_1_6i_16V_Facelift_Tuning_Showcar...   12345678
2897    Escort_MK_1_Hundeknochen_zum_umbauen_auf_RS_2000   11111111
24384                            Schlachte_Golf_3_gt_tdi   11111111
11137  suche_maserati_3200_gt_Zustand_unwichtig_laufe...   10000000
47634                                        Ferrari_FXX    3890000
7814                                         Ferrari_F40    1300000

On the lower end, there are also some odd listings with cars being sold for free.

In [19]:
autos["price_usd"].value_counts().sort_index(ascending=True).head(10)

0     1421
1      156
2        3
3        1
5        2
8        1
9        1
10       7
11       2
12       3
Name: price_usd, dtype: int64

In [20]:
autos_low = autos[autos["price_usd"] < 10]
autos_low[["name", "price_usd"]].sort_values("price_usd", ascending=True).head(10)

name  price_usd
27                Hat_einer_Ahnung_mit_Ford_Galaxy_HILFE          0
32072                                   Suche_3_Golf_GTi          0
32042      VERSCHENKE_DEN_REST_DES_BMW_e39_mit_Kfz_Brief          0
32029  Achtung_Tausche_BMW_E36_Automatic_1_8_Nur_Taus...          0
32024                                         Opel_Corsa          0
31924                                        VW_Polo_G40          0
31911                  Golf_3_1_8_90_pls_TÜV_bis_06/2017          0
31876              Passat_1.8_Turbo___ATM_chip_ca._185ps          0
31858        Opel_corsa_B__ECO__1.0_tauschen_gegen_handy          0
31819                             Audi_A6_quattro_2.5tdi          0

Since this is an auction site, we should expect to see a lot of listings with a starting price at zero. These listings might be probably crawled right after they were listed and had no bidding yet, hence the zero price.

Since we cannot know the settled prices of these listings, we think it's appropriate to remove them and limit our analysis to a price range of 500 - 100,000 USD. With this, the dataset now consists of 45,058 data points.

In [21]:
autos = autos[autos["price_usd"].between(500,100000)]
autos.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 45058 entries, 0 to 49999
Data columns (total 17 columns):
 #   Column              Non-Null Count  Dtype 
---  ------              --------------  ----- 
 0   date_crawled        45058 non-null  object
 1   name                45058 non-null  object
 2   price_usd           45058 non-null  int64 
 3   abtest              45058 non-null  object
 4   vehicle_type        41384 non-null  object
 5   registration_year   45058 non-null  int64 
 6   gearbox             43247 non-null  object
 7   power_ps            45058 non-null  int64 
 8   model               42998 non-null  object
 9   odometer_km         45058 non-null  int64 
 10  registration_month  45058 non-null  int64 
 11  fuel_type           41862 non-null  object
 12  brand               45058 non-null  object
 13  unrepaired_damage   37543 non-null  object
 14  ad_created          45058 non-null  object
 15  postal_code         45058 non-null  int64 
 16  last_seen           45

**2. odometer_km**

In [22]:
autos["odometer_km"].describe()

count     45058.000000
mean     125364.752097
std       39542.959058
min        5000.000000
25%      100000.000000
50%      150000.000000
75%      150000.000000
max      150000.000000
Name: odometer_km, dtype: float64

In [23]:
autos["odometer_km"].value_counts().sort_index(ascending=False)

150000    28695
125000     4836
100000     2029
90000      1676
80000      1384
70000      1189
60000      1129
50000       994
40000       807
30000       759
20000       719
10000       234
5000        607
Name: odometer_km, dtype: int64

From the descriptive statistics table, we can see that there aren't much of unique values in this column. We believe these options are likely set by eBay and users can only choose the closest value for their cars. The majority of listings are used cars with high mileage.

## **Working with dates**

**1. date_crawled & last_seen**

We decide to clean and analyze these two columns together since they contains the dates and times associating with the crawler's task.

At the first glance, we can see that the three columns have the same format with the first 10 characters denoting date and the rest denoting time.


In [24]:
autos[['date_crawled','ad_created','last_seen']][0:5]

date_crawled           ad_created            last_seen
0  2016-03-26 17:47:46  2016-03-26 00:00:00  2016-04-06 06:45:54
1  2016-04-04 13:38:56  2016-04-04 00:00:00  2016-04-06 14:45:08
2  2016-03-26 18:57:24  2016-03-26 00:00:00  2016-04-06 20:15:37
3  2016-03-12 16:58:10  2016-03-12 00:00:00  2016-03-15 03:16:28
4  2016-04-01 14:38:50  2016-04-01 00:00:00  2016-04-01 14:38:50

We think the provided time is too granular and decide to work with only the date instead. Below is the distribution (in percentage) of the listings by date_crawled.

In [25]:
autos["date_crawled"].str[:10].value_counts(normalize=True, dropna=False).sort_index(ascending=True)

2016-03-05    0.025545
2016-03-06    0.014137
2016-03-07    0.036220
2016-03-08    0.033157
2016-03-09    0.032913
2016-03-10    0.032736
2016-03-11    0.033046
2016-03-12    0.037330
2016-03-13    0.015536
2016-03-14    0.036331
2016-03-15    0.034001
2016-03-16    0.029362
2016-03-17    0.031160
2016-03-18    0.012850
2016-03-19    0.034777
2016-03-20    0.038084
2016-03-21    0.037729
2016-03-22    0.032935
2016-03-23    0.032403
2016-03-24    0.028985
2016-03-25    0.031093
2016-03-26    0.032625
2016-03-27    0.031160
2016-03-28    0.034822
2016-03-29    0.033290
2016-03-30    0.033290
2016-03-31    0.031670
2016-04-01    0.033867
2016-04-02    0.035798
2016-04-03    0.038817
2016-04-04    0.036642
2016-04-05    0.013161
2016-04-06    0.003174
2016-04-07    0.001354
Name: date_crawled, dtype: float64

Given the continous date and the relatively consistent distribution, we can conclude that the data was crawled on a daily basis from March 03, 2016 to April 07, 2016. There are some days that the distributions were noticeably lower than others' (e.g., 2016-03-06, 2016-03-13). We think the crawler took a break during these days.

We also take a look at the distribution table of the last_seen column.

In [26]:
autos["last_seen"].str[:10].value_counts(normalize=True, dropna=False).sort_index(ascending=True)

2016-03-05    0.001087
2016-03-06    0.004172
2016-03-07    0.005216
2016-03-08    0.007013
2016-03-09    0.009454
2016-03-10    0.010298
2016-03-11    0.012051
2016-03-12    0.023925
2016-03-13    0.008877
2016-03-14    0.012295
2016-03-15    0.015691
2016-03-16    0.016157
2016-03-17    0.027698
2016-03-18    0.007390
2016-03-19    0.015425
2016-03-20    0.020418
2016-03-21    0.020662
2016-03-22    0.021261
2016-03-23    0.018399
2016-03-24    0.019508
2016-03-25    0.018598
2016-03-26    0.016423
2016-03-27    0.015447
2016-03-28    0.020529
2016-03-29    0.021372
2016-03-30    0.024169
2016-03-31    0.023459
2016-04-01    0.022882
2016-04-02    0.024879
2016-04-03    0.024946
2016-04-04    0.024324
2016-04-05    0.126481
2016-04-06    0.225354
2016-04-07    0.134138
Name: last_seen, dtype: float64

This column has the same date range as date_crawled but there was an increasing trend of distribution toward the end of the period. However, this does not essentially mean that more cars were sold since this column only tells us the time when the ad was last seen by the crawler. As of now, this column does not provide much insights.

**2. ad_created**

In [27]:
autos["ad_created"].str[:10].value_counts(normalize=True, dropna=False).sort_index(ascending=True)

2015-06-11    0.000022
2015-08-10    0.000022
2015-09-09    0.000022
2015-11-10    0.000022
2015-12-05    0.000022
                ...   
2016-04-03    0.039039
2016-04-04    0.037019
2016-04-05    0.011896
2016-04-06    0.003262
2016-04-07    0.001198
Name: ad_created, Length: 76, dtype: float64

Applying the same analysis method to this column, we can see that the dates were discontinous in 2015 and early 2016, and the distribution is minimal as compared to late 2016. A possible explanation is that more people are selling cars and using eBay to advertise their sales.

**3. registration_year & registration_month**

As mentioned in the previous section, this column has some suspicous data that requires our attention. We first look into its decriptive statistics.

In [28]:
autos["registration_year"].describe()

count    45058.000000
mean      2005.063918
std         89.689852
min       1000.000000
25%       2000.000000
50%       2004.000000
75%       2008.000000
max       9999.000000
Name: registration_year, dtype: float64

The min and max values do not make sense since car was not invented in 1000 and cannot be registered in the future. We decide to take a closer look into this column data and display the top and bottom results.

In [29]:
print(autos["registration_year"].value_counts().sort_index(ascending=True).head(5))
print(autos["registration_year"].value_counts().sort_index(ascending=True).tail(5))
autos[~(autos["registration_year"].between(1910,2016))]["registration_year"].value_counts().sum()

1000    1
1001    1
1910    2
1927    1
1929    1
Name: registration_year, dtype: int64
5911    1
6200    1
8888    1
9000    1
9999    3
Name: registration_year, dtype: int64


1774

There are 1,774 suspicous inputs in this column. Since car was invented in the 1880s, we think an acceptable range for registration year should be from 1910 to 2016. The dataset now contains 43,284 data points.

In [30]:
autos = autos[autos["registration_year"].between(1910,2016)]
print(autos["registration_year"].value_counts(normalize=True).sort_values(ascending=False).head(20))
print('\n')
print("Total from 1998 - 2009: ", round(autos["registration_year"].value_counts(normalize=True).sort_values(ascending=False).head(11).sum() * 100, 1),"%")
print('\n')
print(autos.info())

2005    0.066168
2000    0.062702
2004    0.061986
2006    0.061593
2003    0.061385
1999    0.059421
2001    0.058197
2002    0.055887
2007    0.052421
2008    0.050920
2009    0.048009
1998    0.046738
2011    0.037335
2010    0.036549
1997    0.034909
2012    0.030127
1996    0.024767
2016    0.021809
1995    0.019846
2013    0.018321
Name: registration_year, dtype: float64


Total from 1998 - 2009:  63.9 %


<class 'pandas.core.frame.DataFrame'>
Int64Index: 43284 entries, 0 to 49999
Data columns (total 17 columns):
 #   Column              Non-Null Count  Dtype 
---  ------              --------------  ----- 
 0   date_crawled        43284 non-null  object
 1   name                43284 non-null  object
 2   price_usd           43284 non-null  int64 
 3   abtest              43284 non-null  object
 4   vehicle_type        41382 non-null  object
 5   registration_year   43284 non-null  int64 
 6   gearbox             41683 non-null  object
 7   power_ps            43284 non-null  in

Most cars being sold on eBay have the regristration years between 1998 and 2009, accounting for nearly 64% of the listings.

The column registration_month also show some issues since there are 2,972 month values coded as 0. While this seems ood since month values can only range from 1 to 12, upon reading a few listings with a registration month of zero, we think the 0 value simply means that the users forgot the registration month. Therefore, we decide NOT to remove these listings.

In [31]:
autos["registration_month"].value_counts().sort_index(ascending=True)

0     2972
1     2892
2     2686
3     4534
4     3668
5     3667
6     3895
7     3525
8     2850
9     3101
10    3321
11    3093
12    3080
Name: registration_month, dtype: int64

In [32]:
autos[autos["registration_month"] == 0].head(5)

date_crawled  ...            last_seen
46  2016-03-31 10:53:28  ...  2016-04-06 03:44:41
52  2016-03-25 18:50:03  ...  2016-04-07 00:46:00
59  2016-03-17 17:50:54  ...  2016-03-19 17:46:47
60  2016-03-23 21:55:29  ...  2016-04-06 05:45:36
79  2016-03-28 19:58:42  ...  2016-04-07 02:17:59

[5 rows x 17 columns]

## **Working with brands**

Based on the value counts table below, we can see that most of the listings are German brand cars. This makes sense since we are looking at the dataset from eBay German website.

In [33]:
autos["brand"].value_counts(normalize=True)

volkswagen        0.212088
bmw               0.115054
mercedes_benz     0.102278
opel              0.100430
audi              0.090565
ford              0.065197
renault           0.043942
peugeot           0.029387
fiat              0.023542
seat              0.017905
skoda             0.017166
smart             0.015133
mazda             0.015017
nissan            0.015017
citroen           0.014162
toyota            0.013423
hyundai           0.010165
sonstige_autos    0.009588
volvo             0.009380
mini              0.009380
honda             0.007809
mitsubishi        0.007740
kia               0.007347
alfa_romeo        0.006631
chevrolet         0.006030
suzuki            0.005914
porsche           0.005822
chrysler          0.003604
dacia             0.002842
jeep              0.002449
land_rover        0.002264
daihatsu          0.002172
subaru            0.001987
saab              0.001663
jaguar            0.001617
daewoo            0.001340
rover             0.001248
t

Given that the dataset has 40 brands, it is time consuming to conduct an analysis for each brand, we decide to only go for the top common brands that have a distribution more than 5%. This essentially narrows down the brand list from 40 to 6 brands:
- Volkswagen
- BMW
- Mercedes Benz
- Opel
- Audi
- Ford

In [34]:
print("Shares of top six brands: ", round(autos["brand"].value_counts(normalize=True).head(6).sum() * 100,1), "%")
brand_counts = autos["brand"].value_counts(normalize=True)
top_six_brands = brand_counts[brand_counts > 0.05].index
print(top_six_brands)

Shares of top six brands:  68.6 %
Index(['volkswagen', 'bmw', 'mercedes_benz', 'opel', 'audi', 'ford'], dtype='object')


We aggregate the brand column with price_usd and odometer_km to have a better look into these brands regarding price and mileage.

In [35]:
autos_mean_price = {}

for brand in top_six_brands:
    brand_unique = autos[autos["brand"] == brand]
    mean_price = brand_unique["price_usd"].mean()
    autos_mean_price[brand] = int(mean_price)

autos_mean_price_pd = pd.Series(autos_mean_price).sort_values(ascending=False)

autos_mean_price_pd


audi             9571
mercedes_benz    8666
bmw              8447
volkswagen       5783
ford             4247
opel             3394
dtype: int64

In [36]:
autos_mean_mileage = {}

for brand in top_six_brands:
    brand_unique = autos[autos["brand"] == brand]
    mean_mileage = brand_unique["odometer_km"].mean()
    autos_mean_mileage[brand] = int(mean_mileage)

autos_mean_mileage_pd = pd.Series(autos_mean_mileage).sort_values(ascending=False)

autos_mean_price_pd


audi             9571
mercedes_benz    8666
bmw              8447
volkswagen       5783
ford             4247
opel             3394
dtype: int64

In [37]:
top_six_agg = pd.DataFrame(autos_mean_price_pd, columns=["mean_price"])
top_six_agg["mean_mileage"] = autos_mean_mileage_pd

top_six_agg

mean_price  mean_mileage
audi                 9571        128941
mercedes_benz        8666        131083
bmw                  8447        132928
volkswagen           5783        128234
ford                 4247        123520
opel                 3394        128012

The six brands above can be classified into two groups based on their prices:
- Affordable brands: Opel, Ford, and Volkswagen
- Luxury brandsL Audi, Mercedes Benz, BMW

Among each group, we can see that, in general, with the exception of Volswagen, cars with high mileage tend to have low prices.

Between each group, we can also conclude that brand position is a significant factor that determins price: depsite of mileage, Luxury brands often demand high prices.

## **Some futher cleaning**
**1. German to English**

As mentioned in our initial obervations, some columns have texts in German instead of English. We decide to translate and map the values to their English counterparts.

First, we take a look at the dataset again to identify such columns.

In [38]:
autos.head()

date_crawled  ...            last_seen
0  2016-03-26 17:47:46  ...  2016-04-06 06:45:54
1  2016-04-04 13:38:56  ...  2016-04-06 14:45:08
2  2016-03-26 18:57:24  ...  2016-04-06 20:15:37
3  2016-03-12 16:58:10  ...  2016-03-15 03:16:28
4  2016-04-01 14:38:50  ...  2016-04-01 14:38:50

[5 rows x 17 columns]

It seems columns that contain German words are:
- vehicle_type
- gearbox
- fuel_type
- unrepaired_damage

We display the unique values of each column to identify the German words needed to be translated.

In [39]:
# Before translation
print(autos["vehicle_type"].unique())
print(autos["gearbox"].unique())
print(autos["fuel_type"].unique())
print(autos["unrepaired_damage"].unique())

['bus' 'limousine' 'kleinwagen' 'kombi' 'coupe' 'suv' 'cabrio' nan
 'andere']
['manuell' 'automatik' nan]
['lpg' 'benzin' 'diesel' nan 'cng' 'hybrid' 'elektro' 'andere']
['nein' nan 'ja']


In [40]:
german_to_english = {'bus':'bus',
                     'limousine':'limousine',
                     'kleinwagen':'supermini',
                     'kombi':'van',
                     'coupe':'coupe',
                     'suv':'suv',
                     'cabrio':'convertible',
                     'andere':'other',
                     'manuell':'manual',
                     'automatik':'automatic',
                     'lpg':'lpg',
                     'benzin':'petrol',
                     'diesel':'diesel',
                     'cng':'cng',
                     'hybrid':'hybrid',
                     'elektro':'electric',
                     'nein':'no',
                     'ja':'yes'}
for column in ["vehicle_type", "gearbox", "fuel_type", "unrepaired_damage"]:
  autos[column] = autos[column].map(german_to_english)

In [41]:
# After translation
print(autos["vehicle_type"].unique())
print(autos["gearbox"].unique())
print(autos["fuel_type"].unique())
print(autos["unrepaired_damage"].unique())

['bus' 'limousine' 'supermini' 'van' 'coupe' 'suv' 'convertible' nan
 'other']
['manual' 'automatic' nan]
['lpg' 'petrol' 'diesel' nan 'cng' 'hybrid' 'electric' 'other']
['no' nan 'yes']


**2. Unify the dates**

So far, the columns we have worked with that contain timestamps are: date_crawled, last_seen, and ad_created. In this task, we will extract the dates and convert them to be uniform numeric data (e.g., "2016-03-21" becomes the integer 20160321).

When we worked with these columns, we noticed that they have the same format with the first 10 characters denoting dates. Using this observation, we execute the task as following.

In [42]:
for column in ["date_crawled", "last_seen", "ad_created"]:
  autos[column] = autos[column].str[:10].str.replace("-", "").astype(int)

autos.head()

date_crawled  ... last_seen
0      20160326  ...  20160406
1      20160404  ...  20160406
2      20160326  ...  20160406
3      20160312  ...  20160315
4      20160401  ...  20160401

[5 rows x 17 columns]

## **Most common brand & model combinations**

When working with brands, we have identifed the top six common brands listed on eBay German, they are:
- Volkswagen
- BMW
- Mercedes Benz
- Opel
- Audi
- Ford

In this section, we want to dig deeper into these brands and identify the most common model of each brand.



In [43]:
brands = ["audi", "mercedes_benz", "bmw", "volkswagen", "ford", "opel"]

for brand in brands:
  print(brand)
  print(autos[autos["brand"] == brand]["model"].value_counts().sort_values(ascending=False).head(1))
  print('\n')

audi
a4    1199
Name: model, dtype: int64


mercedes_benz
c_klasse    1122
Name: model, dtype: int64


bmw
3er    2519
Name: model, dtype: int64


volkswagen
golf    3437
Name: model, dtype: int64


ford
focus    715
Name: model, dtype: int64


opel
corsa    1266
Name: model, dtype: int64




From the above results, we can see that the most common Brand & Model combinations are:

- Audi A4
- Mercedes Benz C-Class
- BMW 3er
- Volkswagen Golf
- Ford Focus
- Opel Corsa

## **How Damages affect Price**

Wtih this analysis, we want to see how much cars with damages are cheaper than undamaged cars. It seems that undamaged cars are **~4,300 USD** more expensive than damaged cars.

In [53]:
damage = ["yes", "no"]

for answer in damage:
  print(answer)
  print(round(autos[autos["unrepaired_damage"] == answer]["price_usd"].mean(),0))

yes
2779.0
no
7156.0


## **Conclusion**

In this project, we have conducted several cleaning processes and high-level analyses. Some findings and insights that we think worth noticing:
- More cars (both luxury and affordable) are getting listed on eBay;
- Most cars being sold are 7 - 17 years old;
- Cars with high mileage are cheaper;
- The most common brand / model combinations on eBay are: Audi A4, Mercedes Benz C-Class, BMW 3er, Volkswagen Golf, Ford Focus, and Opel Corsa;
- Damaged cars are about 4,300 USD cheaper than undamaged cars.
